### PR TITLE
feat: Expand enumerable types (queue, socket)

### LINKED ISSUE #6: Expand Concat capability, and add Input pipe/queue
Required for https://github.com/sourcenetwork/defradb/issues/1448

### PR DESCRIPTION
**Pull request by AndrewSisley**
Resolves #6 

Expands enumerable types, adding a ring-buffer queue, and a socket, as well as allowing the Concat enumerable to receive new sources after enumeration has begun.

These are to be used by the lens-in-defra code. 

You can see how they are to be used in the WIP draft here: https://github.com/sourcenetwork/defradb/pull/1564 - the code in `Next` function in `lens/lens.go` is the most relevant, and whilst still fairly unpolished it should be fairly stable.

### CODE REVIEW

File: enumerable/concat.go
Comment (AndrewSisley) [inline enumerable/concat.go:41]: thought: I messed up this logic briefly and there is a strong case for adding `concat` tests within this PR I think as this type is now a bit more complicated.  It is tested by defra however.  Let me know, otherwise I will just see how I feel once approved.

Comment (fredcarle) [reply enumerable/concat.go:41]: It would be nice to have it tested within the repo.

File: enumerable/concat.go
Comment (fredcarle) [inline enumerable/concat.go:23]: suggestion: Not sure if I mentioned this before but we should return concrete type when possible. This way type methods don't need to be strictly tied to the interface.

File: enumerable/queue.go
Comment (fredcarle) [inline enumerable/queue.go:50]: suggestion: Same as above. Return concrete type instead of interface.

File: enumerable/socket.go
Comment (fredcarle) [inline enumerable/socket.go:27]: suggestion: Same as above. Return concrete type.

File: enumerable/socket.go
Comment (fredcarle) [inline enumerable/socket.go:7]: question: Can you explain why you decided to name this type `Socket`? At the moment it doesn't jibe with my current definition of a socket.

**Review by fredcarle — CHANGES_REQUESTED** [review]
I have a few suggestions related to return types and a question.

File: enumerable/concat.go
Comment (jsimnz) [inline enumerable/concat.go:35]: question: Little confused on the balance of having both `Append` and `Concat`. Where `Concat` will take a list of sources, append them together, and produce a `Concatentation[T]` type, which also allows additional sources to be appeneded together via `Append` method.

Would it not be better to have a single `Concat` or `Append` top level function with no methods? This would produce a pattern similar to the existing native `append` in go, where you always use the append, and the first parameter of the append is the target slice, and the rest are elements to be added. Should be easy enough to define a 

`EnumerableSlice` type for example, which still implements the Enumerable interface.

File: enumerable/concat.go
Comment (jsimnz) [reply enumerable/concat.go:41]: I feel like this `Next` logic is a lot more complicated then it needs to be, and a few potential oddities, holding off specifics as you mentioned it was briefly messed up, is the current version the *corrected* version.

To make sure we're on the same page on the intended behavior of `Concatenation` types, can you add a docstring to the interface describing the behavior of the new one. I think its straight forward, but just want to confirm, and make sure others that come across this also know.

File: enumerable/queue.go
Comment (jsimnz) [inline enumerable/queue.go:49]: question: Why does a FIFO queue need to be implemented as a ring-buffer? Does this imply you can lose elements in the queue?

File: enumerable/queue.go
Comment (jsimnz) [inline enumerable/queue.go]: I don't see why this is a preferred method for growing the size of the `values` slice. Why not just `append` it? This current method is extremely ineffecient, as it allocates and copys the whole slices every time (short of the ring buffer loop back, but I have a seperate comment above regarding the ring-buffer design).

The native `append` is pretty optimized to avoid unnecessary allocations/copies.

File: enumerable/queue.go
Comment (jsimnz) [inline enumerable/queue.go:39]: question: This seems to be a config/behavior flag, can you document it :)

File: enumerable/queue_test.go
Comment (jsimnz) [inline enumerable/queue_test.go:1]: praise: Great tests 👍

File: enumerable/socket.go
Comment (jsimnz) [inline enumerable/socket.go:50]: nitpick: Slight preference for `var zero T` when using the default/zero value in a generic function like this. (SMALL PREFERENCE, feel free to ignore :) )

File: enumerable/socket.go
Comment (jsimnz) [inline enumerable/socket.go:7]: question: Can you describe the usecase for this particular type?

**Review by jsimnz — CHANGES_REQUESTED** [review]
Have some questions on some of the design choices, no immedage "todos" at the moment, depending on the conversations, there may be followup suggestions/todos.

Marking as "Request Changes" for now just to prevent merge (sorry).

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:41]: > I feel like this Next logic is a lot more complicated then it needs to be

The complication involved is to preserve the order of items, so that Next does not bounce in and out of sources.  We want to yield items from source A, then B, then C, then check if A has new items. 

> To make sure we're on the same page on the intended behavior of Concatenation types, can you add a docstring to the interface describing the behavior of the new one.

I'm not sure I understand you here.  There is only one concrete Concatenation type, and one interface, and they are both doucmented.

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:23]: The concrete types are all private, and the new functions public.

File: enumerable/socket.go
Comment (AndrewSisley) [reply enumerable/socket.go:7]: A socket is a fixture into which other stuff may slot into.

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:35]: > Would it not be better to have a single Concat or Append top level function with no methods?

I am not sure I understand what you mean here.  `Concat` is the constructor, `Append` allows adding new source in to an existing item.

Append needs to live on the concat type, as only a concat type can handle this, and the stuff required for implementation are internal/private.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:49]: It is a dynamically sized ring buffer, no items can be lost.  It is just a more efficienct way of implementing a queue.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:39]: It is not, it is internal stuff required for managing internal state.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go]: The +1 is temporary-ish and is likely to change.  `copy` is much better to use than `append` when adding more than 1 item at a time.

File: enumerable/socket.go
Comment (AndrewSisley) [reply enumerable/socket.go:7]: You can see this being used here: https://github.com/sourcenetwork/defradb/pull/1564/files#diff-e7f1317923231a3017c3a580d21ce39346555446638d045a69310b1a1aeab15d in the MigrateUp function.  It allows  enumerables to be defined before their source, and for sources to be swapped in and out.

File: enumerable/queue.go
Comment (jsimnz) [reply enumerable/queue.go]: copy most certainly is not better when adding more than 1 item. append can append as many items as you want, and will determine the ideal allocations/copies when needed.

Comment (AndrewSisley) [reply enumerable/queue.go]: I think you are mistaken, check the benches in this blog post: https://eremeev.ca/posts/golang-copy-vs-append/

File: enumerable/socket.go
Comment (fredcarle) [reply enumerable/socket.go:7]: I find it confusing as it leads me to believe that it relates to network sockets. Don't you think?

File: enumerable/concat.go
Comment (fredcarle) [reply enumerable/concat.go:23]: You can make the concrete types public then. I really think that unless an interface method needs an interface return type that we try to avoid returning interface types

File: enumerable/socket.go
Comment (AndrewSisley) [reply enumerable/socket.go:7]: network sockets took the name from the same non-computing concept and I see them as being conceptually quite similar, only one is remote/cross-process, and this is not.

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:23]: > You can make the concrete types public then. I really think that unless an interface method needs an interface return type that we try to avoid returning interface types

I am unsure as to how to express just how much I really do not want to make the concrete types public.

Comment (fredcarle) [reply enumerable/concat.go:23]: Why? You don't have to make the fields public. It's much better than returning an interface type.

Comment (AndrewSisley) [reply enumerable/concat.go:23]: Nope.  This are internal types, and I highly value the ability to rip and replace them, or allow for more complex constructors which may not return a single concrete type without having to worry about people directly referencing them.

These are not run of the mill business-logic types, but super generic data structures, and I am somewhat sceptical of anyone's ability to convince me to expose them in the short-medium term.  I think that is very bad thing to do.

There is absolutely no reason for anyone to need to directly interact with the concrete types in this package.

Comment (jsimnz) [reply enumerable/concat.go:23]: I like the interfaces (granted I'm a little interface happy compared to average Go dev 😉)

Comment (fredcarle) [reply enumerable/concat.go:23]: > Nope. This are internal types, and I highly value the ability to rip and replace them, or allow for more complex constructors which may not return a single concrete type without having to worry about people directly referencing them.

Returning a concrete type makes zero difference for this statement. You can still rip and replace without worrying about people directly referencing them. 

> There is absolutely no reason for anyone to need to directly interact with the concrete types in this package.

Again, returning a concrete type makes zero difference  for this statement. All the methods relate to the concrete type and if there are no public fields the liming factor is the same. 

The drawback with returning an interface is that if you want to add functionality to the type and still be able to use the constructor, you have to also change the interface which one might not want to do. It's much more flexible.

Comment (fredcarle) [reply enumerable/concat.go:23]: > I like the interfaces (granted I'm a little interface happy compared to average Go dev 😉)

The number of times I've hit walls because of us using interfaces for return types 😤... lol

Comment (fredcarle) [reply enumerable/concat.go:23]: For reference: https://github.com/golang/go/wiki/CodeReviewComments#interfaces

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:41]: Tests have been added

File: enumerable/concat.go
Comment (islamaliev) [inline enumerable/concat.go:5]: suggestion: I think the name does not quite reflect it's purpose.
If it's `Enumerable` that can be expanded why not to call it `ExpendableEnumerable` or `DynamicEnumerable`

File: enumerable/concat.go
Comment (islamaliev) [inline enumerable/concat.go:45]: suggestion: I think it's worth mentioning that this scenario is only possible if `Next` is called on an exhausted enumerable.

File: enumerable/concat_test.go
Comment (islamaliev) [inline enumerable/concat_test.go:51]: suggestion: these repetitive checks can be put in a loop

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go:16]: question: why don't we have `Size` for `Enumerable`?

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go]: todo: this is not a rate. Rate should be a floating number. The better name would be `growthSize`

File: enumerable/queue.go
Comment (islamaliev) [reply enumerable/queue.go:39]: documentation would be helpful

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go]: todo: why to increment capacity by 1?

This is extremely inefficient for any reasonably sized data structure.
I'd say we need to go with 1.5 ratio. Can't think of any disadvantages

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go:36]: suggestion: wouldn't it be more intuitive to call these two `readIndex` and `writeIndex`?

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go:80]: question: why not rearrange the elements?

I see now the reason why you grow the size by 1: it's easier to handle this scenario.
I would say when this happens we can rearrange from:
```
[3,4,1,2]
     ^
     |
     read and write index
```
to this:
```
[1,2,3,4,_,_,_]
 ^       ^
 |       |
 |       write index
 read index
```
The queue is not thread-safe anyway

File: enumerable/concat_test.go
Comment (islamaliev) [inline enumerable/concat_test.go:9]: suggestion: may try the agreed upon test name structure: `<category>_<condition>_<result>`?

This example would be `TestConcat_IfEmpty_YieldNothing`

File: enumerable/queue_test.go
Comment (islamaliev) [inline enumerable/queue_test.go]: praise: helpful visualisation

**Review by islamaliev — CHANGES_REQUESTED** [review]
in general looks good, but there are some major concerns related to performance.

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:5]: `expanded` or `Dynamic` is way too vague for my liking as it can mean a whole bunch of stuff.  And concat/concatenation is a fairly common term for this behaviour.

File: enumerable/concat.go
Comment (AndrewSisley) [reply enumerable/concat.go:45]: Not a problem, will add

- [x] doc if case

File: enumerable/concat_test.go
Comment (AndrewSisley) [reply enumerable/concat_test.go:51]: I kind of dislike that.  We are testing a small number of specific items, and in some cases, they are from different sources. I see a loop as a complication and abstraction that detracts from the readability of the test.

I also thought about refactoring it out into a private test func, but that also masks the Act and Assert part of the test and I much prefer them to be highly visible.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:16]: Would mean complicating the core interface with a function that is not used in production code.  And implementing it on every type, regardless as to whether it makes any sense on that type (I dont think it does for a lot of them, many just have a `source` enumerable and a predicate/int/etc).

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go]: Agreed, and I felt a little uncomfortable calling it rate initially but failed to come up with an alternative at the time.  Size is much more accurate, thanks.

- [x] Rename const

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go]: This is documented on the const.

File: enumerable/queue.go
Comment (fredcarle) [reply enumerable/queue.go:16]: What you just stated Andy is one of the reasons why returning the concrete type is better practice in Go. If you want a `Size` when it make sense you don't need to modify the interface for it :)

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:36]: I think read/write is less accurate though, especially given that the two props actually differ in terms of temporality: `currentIndex` *is* the last index produced by a call to `Next`, that may or may not be out of bounds, and may or may not have been read via `Value`.  `lastSetIndex` *has* been set, was valid, and is essentially a record of the past.

Writing all that out though, I may have missed something whilst testing.

- [x] Test post-loop Next=>Put=>Value - I'm pretty sure it is broken, and that a space between these two integers should always been maintained to prevent the premature overwrite of `Value`

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:80]: That is an interesting thought. Will look at this some more. I do worry that this might make an future threadsafetiness harder/more-expensive to achieve, but maybe not really anymore than allowing the buffer to grow already does.

What benefit do you see in this though?

File: enumerable/concat_test.go
Comment (AndrewSisley) [reply enumerable/concat_test.go:9]: I had already added tests before that conversation took place, and I'd prefer not to mix and match within a PR, and I'd also rather not rename everything.  The agreement was more of a 'lets try this for a bit' agreement, not a 'lets convert the entire codebase right now' to this agreement.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:39]: Will add, although I'm not sure much can be added to the declarations themselves.

- [x] doc private internal state props

File: enumerable/queue.go
Comment (islamaliev) [reply enumerable/queue.go:80]: I though this was the main reason why you chose (for now) to grow the array by 1. So that it's easier to manage indexes. Otherwise your read index will have to keep in mind that there are empty slots for future writing that it will have to jump over. 
```
[1,2,_,_,_,3,4]
   ^ ^     ^
   | |     next read index
   | write index
   read index
 ```
But with the rearrangement you can grow it as much as you want.

File: enumerable/queue.go
Comment (islamaliev) [inline enumerable/queue.go:21]: question: I can't see why we would want to change it in the future and not now. (original comment https://github.com/sourcenetwork/immutable/pull/7#discussion_r1236715588)

Would be nice to hear some benefits of growing the size by 1.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:80]: Ah no, the +1 is because I only want it to grow by one.  The problem you just described doesnt exist, as read never crosses the gap (it is a queue, if an item is read it is consumed and never gets read again - if the read index reaches the gap it means the queue is currently empty).

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:21]: Because +1 fits the current use case very well, but in the future that may change as both the scope of Lens-in-Defra grows, and/or if other areas wish to use this type.

Please remember that Lens in Defra is currently as simple as it will get, and it will very likely continue to grow in complexity.

File: enumerable/queue.go
Comment (AndrewSisley) [reply enumerable/queue.go:80]: Note: I actually tried using this setup last week to help solve an edge case.  It did not simplify the code unfortunately.

**Review by fredcarle — APPROVED** [review]
LGTM

File: enumerable/queue.go
Comment (islamaliev) [reply enumerable/queue.go:21]: I don't think it's a valid argument when developing an independent component, which the whole `enumerable` is.
We should keep in mind ideally all or at least major use cases, and not only specific ones.
What if we decide (which is highly likely) to use the queue in some other component and stuff it with thousands of items?
The cost of making it more future-proof is low, so I still don't understand why we don't do it now.

Comment (fredcarle) [reply enumerable/queue.go:21]: I agree with you Islam and I'm not a fan of using copy in this situation either but for the sake of moving this PR ahead, I think we can leave it as is and if we end up using it elsewhere, we can make the change at that time.

**Pull request merged**
